File: QUESTIONS/Codevita/ladder.cpp
#include <bits/stdc++.h>
using namespace std;

struct Point {
    int row, col;
    bool operator<(const Point &other) const {
        return row == other.row ? col < other.col : row < other.row;
    }
    bool operator==(const Point &other) const {
        return row == other.row && col == other.col;
    }
};

bool insideGrid(int r, int c, int R, int C) {
    return r >= 0 && r < R && c >= 0 && c < C;
}

bool isValidPlacement(const vector<Point> &cells, const vector<string> &board, int R, int C) {
    for (size_t i = 0; i < cells.size(); i++) {
        const Point &p = cells[i];

        if (!insideGrid(p.row, p.col, R, C) || board[p.row][p.col] == 'B')
            return false;
    }
    return true;
}

bool checkBoundingSquare(const vector<Point> &oldCells, const vector<Point> &newCells) {
    vector<Point> combined = oldCells;
    combined.insert(combined.end(), newCells.begin(), newCells.end());
    int k = oldCells.size();

    vector<int> rows, cols;
    for (size_t i = 0; i < combined.size(); i++) {
        rows.push_back(combined[i].row);
        cols.push_back(combined[i].col);
    }

    int rmin = *min_element(rows.begin(), rows.end());
    int rmax = *max_element(rows.begin(), rows.end());
    int cmin = *min_element(cols.begin(), cols.end());
    int cmax = *max_element(cols.begin(), cols.end());

    return (rmax - rmin == k - 1) && (cmax - cmin == k - 1);
}

vector<vector<Point>> generatePossibleRotations(const vector<Point> &cells, const vector<string> &board, int R, int C) {
    int k = cells.size();
    vector<vector<Point>> results;

    bool horizontal = true;
    for (int i = 1; i < k; i++) {
        if (cells[i].row != cells[0].row) {
            horizontal = false;
            break;
        }
    }

    if (horizontal) {
        for (int r = 0; r <= R - k; r++) {
            vector<Point> temp;
            for (int i = 0; i < k; i++)
                temp.push_back(Point{r + i, cells[0].col});
            if (isValidPlacement(temp, board, R, C) && checkBoundingSquare(cells, temp))
                results.push_back(temp);
        }
    } else {
        for (int c = 0; c <= C - k; c++) {
            vector<Point> temp;
            for (int i = 0; i < k; i++)
                temp.push_back(Point{cells[0].row, c + i});
            if (isValidPlacement(temp, board, R, C) && checkBoundingSquare(cells, temp))
                results.push_back(temp);
        }
    }

    return results;
}

vector<vector<Point>> generateTranslations(const vector<Point> &cells, int R, int C) {
    vector<vector<Point>> translations;
    vector<pair<int, int>> directions;
    directions.push_back(make_pair(-1, 0));
    directions.push_back(make_pair(1, 0));
    directions.push_back(make_pair(0, -1));
    directions.push_back(make_pair(0, 1));

    for (size_t d = 0; d < directions.size(); d++) {
        int dr = directions[d].first;
        int dc = directions[d].second;

        vector<Point> moved;
        for (size_t i = 0; i < cells.size(); i++)
            moved.push_back(Point{cells[i].row + dr, cells[i].col + dc});

        bool ok = true;
        for (size_t i = 0; i < moved.size(); i++)
            if (!insideGrid(moved[i].row, moved[i].col, R, C))
                ok = false;
        if (ok) translations.push_back(moved);
    }

    return translations;
}

int minMoves(vector<string> &board, int R, int C, vector<Point> start, vector<Point> target) {
    sort(start.begin(), start.end());
    sort(target.begin(), target.end());
    if (start == target) return 0;

    queue<pair<vector<Point>, int>> q;
    set<vector<Point>> visited;

    q.push(make_pair(start, 0));
    visited.insert(start);

    while (!q.empty()) {
        pair<vector<Point>, int> front = q.front(); q.pop();
        vector<Point> current = front.first;
        int dist = front.second;

        vector<vector<Point>> trans = generateTranslations(current, R, C);
        for (size_t i = 0; i < trans.size(); i++) {
            vector<Point> next = trans[i];
            if (!visited.count(next) && isValidPlacement(next, board, R, C)) {
                if (next == target) return dist + 1;
                visited.insert(next);
                q.push(make_pair(next, dist + 1));
            }
        }

        vector<vector<Point>> rotations = generatePossibleRotations(current, board, R, C);
        for (size_t i = 0; i < rotations.size(); i++) {
            vector<Point> rot = rotations[i];
            if (!visited.count(rot)) {
                if (rot == target) return dist + 1;
                visited.insert(rot);
                q.push(make_pair(rot, dist + 1));
            }
        }
    }

    return -1;
}

vector<Point> collectCells(vector<string> &board, char ch) {
    vector<Point> res;
    for (size_t r = 0; r < board.size(); r++)
        for (size_t c = 0; c < board[0].size(); c++)
            if (board[r][c] == ch)
                res.push_back(Point{int(r), int(c)});
    return res;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int R, C;
    cin >> R >> C;
    vector<string> board(R);
    for (int i = 0; i < R; i++)
        cin >> board[i];

    vector<Point> start = collectCells(board, 'L');
    vector<Point> target = collectCells(board, 'T');

    int ans = minMoves(board, R, C, start, target);
    if (ans != -1) cout << ans;
    else cout << "Impossible";
    return 0;
}

File: QUESTIONS/Codevita/LADDER.py
from collections import deque

def in_bounds(r, c, M, N):
    return 0 <= r < M and 0 <= c < N

def can_place(cells, grid, M, N):
    for r, c in cells:
        if not in_bounds(r, c, M, N) or grid[r][c] == 'B':
            return False
    return True

def bounding_square_ok(old_cells, new_cells, grid, M, N):
    all_cells = list(old_cells) + list(new_cells)
    rs = [r for r, _ in all_cells]
    cs = [c for _, c in all_cells]
    k = len(old_cells)
    rmin, rmax = min(rs), max(rs)
    cmin, cmax = min(cs), max(cs)
    if (rmax - rmin != k - 1) or (cmax - cmin != k - 1):
        return False
    return True

def generate_rotations(cells, grid, M, N):
    k = len(cells)
    rows = [r for r, _ in cells]
    horizontal = all(r == rows[0] for r in rows)
    rotations = []
    if horizontal:
        for r in range(M - k + 1):
            new = tuple(sorted((r + i, cells[0][1]) for i in range(k)))
            if can_place(new, grid, M, N) and bounding_square_ok(cells, new, grid, M, N):
                rotations.append(new)
    else:
        for c in range(N - k + 1):
            new = tuple(sorted((cells[0][0], c + i) for i in range(k)))
            if can_place(new, grid, M, N) and bounding_square_ok(cells, new, grid, M, N):
                rotations.append(new)
    return rotations

def neighbors_translate(cells, M, N):
    for dr, dc in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
        nxt = tuple(sorted((r + dr, c + dc) for r, c in cells))
        if all(in_bounds(r, c, M, N) for r, c in nxt):
            yield nxt

def min_steps(grid, M, N, start_cells, target_cells):
    start = tuple(sorted(start_cells))
    target = tuple(sorted(target_cells))
    if start == target:
        return 0
    q = deque([(start, 0)])
    visited = {start}
    while q:
        cur, dist = q.popleft()
        for nxt in neighbors_translate(cur, M, N):
            if nxt not in visited and can_place(nxt, grid, M, N):
                if nxt == target:
                    return dist + 1
                visited.add(nxt)
                q.append((nxt, dist + 1))
        for rot in generate_rotations(cur, grid, M, N):
            if rot not in visited:
                if rot == target:
                    return dist + 1
                visited.add(rot)
                q.append((rot, dist + 1))
    return None

def find_cells(grid, ch):
    return [(i, j) for i in range(len(grid)) for j in range(len(grid[0])) if grid[i][j] == ch]

if __name__ == "__main__":
    M, N = map(int, input().split())
    grid = [list(input().strip()) for _ in range(M)]
    start = find_cells(grid, 'L')
    target = find_cells(grid, 'T')
    res = min_steps(grid, M, N, start, target)
    print(res if res is not None else "Impossible", end="")
